MQTTnet.Core/Client/MqttClientQueued.cs

```diff
@@ -0,0 +1,166 @@
+                {
+                    if (_persistentMessagesManager == null)
+                        _persistentMessagesManager = new MqttClientQueuedPersistentMessagesManager(_options);
+                    await _persistentMessagesManager.LoadMessagesAsync();
+                    await InternalPublishAsync(_persistentMessagesManager.GetMessages(), false);
+                }
+            }
+            catch (Exception)
+            {
+                await DisconnectAsync().ConfigureAwait(false);
+                throw;
+            }
+        }
+
+        public new async Task PublishAsync(IEnumerable<MqttApplicationMessage> applicationMessages)
+        {
+            await InternalPublishAsync(applicationMessages, true);
+        }
+
+        private async Task InternalPublishAsync(IEnumerable<MqttApplicationMessage> applicationMessages, bool appendIfUsePersistance)
+        {
+            ThrowIfNotConnected();
+
+            foreach (var applicationMessage in applicationMessages)
+            {
+                if (_usePersistance && appendIfUsePersistance)
+                    await _persistentMessagesManager.SaveMessageAsync(applicationMessage);
+
+                _inflightQueue.Enqueue(applicationMessage);
+            }
+        }
+
+        public new async Task<IList<MqttSubscribeResult>> SubscribeAsync(IEnumerable<TopicFilter> topicFilters)
+        {
+            return await base.SubscribeAsync(topicFilters);
+        }
+
+        private void ThrowIfNotConnected()
+        {
+            if (!IsConnected) throw new MqttCommunicationException("The client is not connected.");
+        }
+
+        private ushort GetNewPacketIdentifier()
+        {
+            return (ushort)Interlocked.Increment(ref _latestPacketIdentifier);
+        }
+        
+
+        private void SetupOutgoingPacketProcessingAsync()
+        {
+#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+            Task.Factory.StartNew(
+                () => SendPackets(base._cancellationTokenSource.Token),
+                base._cancellationTokenSource.Token,
+                TaskCreationOptions.LongRunning,
+                TaskScheduler.Default).ConfigureAwait(false);
+#pragma warning restore CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+        }
+
+        private async Task SendPackets(CancellationToken cancellationToken)
+        {
+            MqttNetTrace.Information(nameof(MqttClient), "Start sending packets.");
+            MqttApplicationMessage messageToSend = null;
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    while (_inflightQueue.TryDequeue(out messageToSend))
+                    {
+                        MqttQualityOfServiceLevel qosGroup = messageToSend.QualityOfServiceLevel;
```

Comment:
Why not just using the "PublishAsync" method and catch all exceptions? In the catch clause I would assume that:
1) QoS 0 will be dropped
2 ) QoS 1 will be dropped
3 ) QoS 2 will be enqueued (or inserted as item 0?)

Reply:
Right. I agree

Reply:
Done, but to be tested more